File: playground/payload/src/collections/Workshops.ts
import { BlocksFeature, lexicalEditor } from '@payloadcms/richtext-lexical'
import type { Block, CollectionConfig, Field } from 'payload'

type WorkshopValues = {
  city?: string | null
  title?: string | null
}

const authenticated = ({ req }: { req: { user: unknown } }) => Boolean(req.user)

const workshopFields = (): Field[] => [
  {
    name: 'city',
    type: 'select',
    required: true,
    options: ['London', 'Bristol'],
    admin: {
      description: 'After editing the title, change this city to refresh its feedback.',
    },
  },
  {
    name: 'title',
    type: 'text',
    label: 'Workshop title',
    required: true,
    validate: (
      value: null | string | undefined,
      { siblingData }: { siblingData: unknown },
    ) => {
      const { city } = siblingData as WorkshopValues

      if (!value?.trim() || !city) {
        return true
      }

      return value.toLocaleLowerCase().includes(city.toLocaleLowerCase())
        ? true
        : `Include ${city} in the workshop title so guests know where to go.`
    },
    admin: {
      description:
        'Include the selected city, for example: London pottery evening. Try removing London, then wait a moment without saving.',
    },
  },
]

const WorkshopCard: Block = {
  slug: 'workshop',
  labels: {
    singular: 'Workshop card',
    plural: 'Workshop cards',
  },
  fields: workshopFields(),
}

export const Workshops: CollectionConfig = {
  slug: 'workshops',
  admin: {
    useAsTitle: 'title',
  },
  access: {
    create: authenticated,
    read: authenticated,
    update: authenticated,
    delete: authenticated,
  },
  fields: [
    ...workshopFields(),
    {
      name: 'sessions',
      type: 'array',
      label: 'Extra sessions',
      fields: workshopFields(),
      admin: {
        description:
          'The same rule works inside each row. Try editing a title and moving that row.',
      },
    },
    {
      name: 'description',
      type: 'richText',
      label: 'Description and embedded workshop card',
      editor: lexicalEditor({
        features: ({ rootFeatures }) => [
          ...rootFeatures,
          BlocksFeature({ blocks: [WorkshopCard] }),
        ],
      }),
      admin: {
        description:
          'Choose Edit on the seeded Workshop card. Its title is checked before you choose Apply.',
      },
    },
  ],
}

File: playground/ridu/content/workshops.go
package content

import (
	"strings"

	"github.com/riducms/ridu"
	"github.com/riducms/ridu/field"
	"github.com/riducms/ridu/operation"
	"github.com/riducms/ridu/plugins/richtext"
)

var Workshops = ridu.Collection{
	Slug: "workshops",
	Admin: ridu.CollectionAdmin{
		UseAsTitle: "title",
	},
	Access: ridu.CollectionAccess{
		Admin:  authenticatedOnly,
		Create: authenticatedOnly,
		Read:   authenticatedOnly,
		Update: authenticatedOnly,
		Delete: authenticatedOnly,
	},
	Fields: append(workshopFields(),
		field.Array("sessions", workshopFields()).
			Label("Extra sessions").
			Admin(field.Admin{
				Description: "The same rule works inside each row. Try editing a title and moving that row.",
			}),
		richtext.Field("description", richtext.Config{
			Blocks: []field.Block{
				{
					Slug:   "workshop",
					Labels: field.BlockLabels{Singular: "Workshop card", Plural: "Workshop cards"},
					Fields: workshopFields(),
				},
			},
		}).
			Label("Description and embedded workshop card").
			Admin(field.Admin{
				Description: "Choose Edit on the seeded Workshop card. Its title is checked before you choose Apply.",
			}),
	),
}

// The same small factory works in a document, repeated rows and rich-text cards.
func workshopFields() field.Fields {
	return field.Fields{
		field.Select("city", "London", "Bristol").
			Required().
			Admin(field.Admin{
				Description: "After editing the title, change this city to refresh its feedback.",
			}),
		field.Text("title").
			Required().
			Label("Workshop title").
			Admin(field.Admin{
				Description: "Include the selected city, for example: London pottery evening. Try removing London, then wait a moment without saving.",
			}).
			Validate(func(
				ctx operation.ValidationContext,
				value operation.Value[string],
			) ([]operation.Issue, error) {
				return checkTitle(ctx.Siblings, value)
			}).
			LiveValidate(func(
				ctx operation.LiveValidationContext,
				value operation.Value[string],
			) ([]operation.Issue, error) {
				return checkTitle(ctx.Siblings, value)
			}),
	}
}

// One Go business rule, used both for advisory feedback and authoritative Save.
func checkTitle(siblings operation.View, value operation.Value[string]) ([]operation.Issue, error) {
	title, present := value.Get()
	city, selected := siblings.String("city")
	if !present || strings.TrimSpace(title) == "" || !selected || city == "" {
		return nil, nil // Required handles empty values when saving.
	}
	if !strings.Contains(strings.ToLower(title), strings.ToLower(city)) {
		return []operation.Issue{
			{
				Code:    "city_in_title",
				Message: "Include " + city + " in the workshop title so guests know where to go.",
			},
		}, nil
	}
	return nil, nil
}
